lib/CombinedCancellationToken.php

```diff
@@ -4,12 +4,9 @@
 
 final class CombinedCancellationToken implements CancellationToken
 {
-    /** @var array{0: CancellationToken, 1: string}[] */
+    /** @var array<string, CancellationToken> */
     private $tokens = [];
 
-    /** @var string */
-    private $nextId = "a";
```

Comment:
There is a problem here: https://github.com/amphp/amp/blob/master/lib/CombinedCancellationToken.php#L33

Tokens stored in array and there could be duplicate $id in value. So if i call CombinedCancellationToken->unsubscribe($id) i can not remove from $this->tokens, because $id is not unique.

Reply:
Also, that is only a problem because you changed how the tokens are stored.

Reply:
I dont understand how to find item by id if i can not change pattern for generated id.

Here passes id like a,b,c,d:
```
    /** @inheritdoc */
    public function unsubscribe(string $id)
    {
        unset($this->callbacks[$id]);
        // how to find object by id if there is multiple tokens with same id?
        // unset($this->tokens[$id]);
    }
```

Reply:
You should not be unsetting `$this->tokens[$id]` in `unsubscribe`. That is the subscription of the combined token to the original tokens, which should not be removed until the combined token is destroyed.

Reply:
Why my example is not destroying  combined token? (look my first message with example loop with requests.

Reply:
The core problem is the circular reference created by the closure on line 22. Make this closure static and use references to the instance variables as @bwoebi suggested. An example of this can be found [here](https://github.com/amphp/byte-stream/blob/10c420782bd805c4af634bab69b5dfdc14ab721c/lib/ResourceInputStream.php#L67-L76).

Reply:
@trowski, thanks. You are right. In my another example, i forgot to extract shared pool and client to constructor.
Now it works like a charm. 👍  Code updated in MR.